### PR TITLE
feat: uniformize inputs and outputs for all algos

### PR DESCRIPTION
**Pull request by louishulot**
Use `datasamples` as input identifiers for X and y accros all functions. 
Pass `rank` within a new argument named `task_properties`

Companion PRs: 
- substra: https://github.com/Substra/substra/pull/284
- substrafl: https://github.com/Substra/substrafl/pull/20
- substra-tests: https://github.com/Substra/substra-tests/pull/207
- substra-tools (main) https://github.com/Substra/substra-tools/pull/58
- substra-documentation:  (here)  https://github.com/Substra/substra-documentation/pull/179

### CODE REVIEW

File: substrafl_examples/strategies_examples/plot_substrafl_fedavg.py [hunk 1]
```diff
@@ -403,9 +403,9 @@ def forward(self, x, eval=False):
 
 
 class TorchDataset(torch.utils.data.Dataset):
-    def __init__(self, x, y, is_inference: bool):
-        self.x = x
-        self.y = y
+    def __init__(self, datasamples, is_inference: bool):
+        self.x = torch.FloatTensor(datasamples["images"][:, None, ...])
+        self.y = F.one_hot(torch.from_numpy(datasamples["labels"]).type(torch.int64), 10).type(torch.float32)
```
Comment (RomainGoussault) [inline substrafl_examples/strategies_examples/plot_substrafl_fedavg.py:408]: Why do we have to convert twice here?

File: docs/source/conf.py [hunk 2]
```diff
@@ -27,10 +27,11 @@
...
     SUBSTRAFL_VERSION = "0.26.0"
 
+# TODO: change before merge (to build in local)
 else:
-    SUBSTRA_VERSION = "main"
-    TOOLS_VERSION = "main"
-    SUBSTRAFL_VERSION = "main"
+    SUBSTRA_VERSION = "gt/uniformize_inputs"
+    TOOLS_VERSION = "gt/uniformize_inputs"
+    SUBSTRAFL_VERSION = "gt/uniformize_inputs"
```
Comment (Esadruhn) [inline docs/source/conf.py:34]: TODO

File: substrafl_examples/strategies_examples/assets/dataset/description.md [hunk 3]
```diff
@@ -14,9 +14,5 @@ The target is the number (0 -> 9) represented by the pixels.
 
 The opener exposes 6 methods:
 
-* `get_X` returns a torch.FloatTensor containing the images
-* `get_y` returns a torch.FloatTensor containing the labels converted in OneHot tensor
-* `fake_X` returns a fake data sample of images
-* `fake_y`returns a fake data sample of labels
-* `save_predictions` saves a numpy array as npy,
-* `get_predictions` loads the npy saved with `save_predictions` and returns a numpy array
+- `get_data` returns a torch.FloatTensor containing the images and the labels in a dict
```
Comment (Esadruhn) [inline substrafl_examples/strategies_examples/assets/dataset/description.md:17]:
```suggestion
- `get_data` returns a dictionary containing containing the images and the labels as numpy arrays
```

File: substrafl_examples/strategies_examples/plot_substrafl_fedavg.py [hunk 4]
```diff
@@ -62,6 +62,10 @@
 
 N_CLIENTS = 2
 
+# Choose the subprocess mode to locally simulate the FL process
```
Comment (Esadruhn) [inline substrafl_examples/strategies_examples/plot_substrafl_fedavg.py:65]: please do not update this, it is already fixed in another PR: https://github.com/Substra/substra-documentation/pull/178

File: substrafl_examples/strategies_examples/plot_substrafl_fedavg.py [hunk 5]
```diff
@@ -391,7 +391,7 @@ def forward(self, x, eval=False):
 # dataset to be stateful.
 
 # Number of model update between each FL strategy aggregation.
-NUM_UPDATES = 15
+NUM_UPDATES = 100
```
Comment (Esadruhn) [inline substrafl_examples/strategies_examples/plot_substrafl_fedavg.py:394]:
```suggestion
NUM_UPDATES = 15
```

Is there a reason to change this parameter?

**Review by Esadruhn — CHANGES_REQUESTED** [review]

File: substrafl_examples/strategies_examples/plot_substrafl_fedavg.py (hunk 1 above)
Comment (louishulot) [reply substrafl_examples/strategies_examples/plot_substrafl_fedavg.py:408]: I guess, int for one hot encoding then float for the network (I took the code from the example so I can't say much)

File: substrafl_examples/strategies_examples/plot_substrafl_fedavg.py (hunk 5 above)
Comment (louishulot) [reply substrafl_examples/strategies_examples/plot_substrafl_fedavg.py:394]: Having a performance of 0.6 for Mnsit. This increases the AUC to 0.85 (still bad but better)

File: substrafl_examples/strategies_examples/assets/dataset/description.md (hunk 3 above)
Comment (louishulot) [reply substrafl_examples/strategies_examples/assets/dataset/description.md:17]: done

File: examples/titanic_example/assets/algo_random_forest/titanic_algo_rf.py [hunk 6]
```diff
@@ -117,10 +117,10 @@ def _predict_pandas(self, model, X):
         y_pred = model.predict(X)
         return pd.DataFrame(columns=["Survived"], data=y_pred)
 
-    def train(self, inputs, outputs):
+    def train(self, inputs, outputs, task_properties):
 
-        X = inputs["X"]
-        y = inputs["y"]
+        X: pd.DataFrame = inputs["datasamples"].drop(columns="Survived")
+        y: pd.Series = inputs["datasamples"].Survived
```
[the remark this answers is not in the document]
Comment (louishulot) [reply examples/titanic_example/assets/algo_random_forest/titanic_algo_rf.py:123]: done

File: substrafl_examples/strategies_examples/assets/dataset/description.md (hunk 3 above)
Comment (louishulot) [reply substrafl_examples/strategies_examples/assets/dataset/description.md:17]: done

File: substrafl_examples/strategies_examples/plot_substrafl_fedavg.py (hunk 4 above)
Comment (louishulot) [reply substrafl_examples/strategies_examples/plot_substrafl_fedavg.py:65]: done

File: examples/titanic_example/assets/algo_random_forest/titanic_algo_rf.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (louishulot) [reply examples/titanic_example/assets/algo_random_forest/titanic_algo_rf.py:123]: done

**Review by Esadruhn — APPROVED** [review]

**Pull request merged**